{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "280e8954-6fc7-4399-94e6-a69598f20296",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tensorflow as tf\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.neighbors import KernelDensity\n",
    "from sklearn.metrics import r2_score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0a5b1d8-f42b-4aea-abcb-e00a33fc4d65",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"mTRAQ_WC_THP1_mac_library_04202023.tsv\"\n",
    "df = pd.read_csv(path, delimiter='\\t')\n",
    "original_df = pd.DataFrame(df)\n",
    "distinct_df = original_df.drop_duplicates(subset='transition_group_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdf6febf-5a9b-4c48-a7db-a486ebf069e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide_sequence = \"ACDEFGHIKLMNPQRSTVWYacdefghiklmnpqrstvwy\"   # define peptide sequence as a string \n",
    "max_sequence_length = 30    # set the maximum sequence length\n",
    "num_amino_acids = 20      # number of possible amino acids\n",
    "\n",
    "# create a dictionary to map amino acids to their indices\n",
    "amino_acid_to_index = {aa: i for i, aa in enumerate(peptide_sequence)}\n",
    "\n",
    "def one_hot_encode_sequence(sequence):\n",
    "    one_hot_sequence = np.zeros((max_sequence_length, num_amino_acids), dtype=np.float32)\n",
    "    for i, aa in enumerate(sequence[:max_sequence_length]):\n",
    "        aa_index = amino_acid_to_index.get(aa.upper(), None)\n",
    "        if aa_index is not None:\n",
    "            one_hot_sequence[i, aa_index] = 1.0\n",
    "    return one_hot_sequence\n",
    "\n",
    "X_peptide = [one_hot_encode_sequence(seq) for seq in distinct_df['PeptideSequence']]\n",
    "\n",
    "# One-hot encode charge states (assumes a maximum charge state of 5)\n",
    "max_charge_state = 5\n",
    "X_charge = np.zeros((len(distinct_df), max_charge_state), dtype=np.float32)\n",
    "\n",
    "for i, charge in enumerate(distinct_df['PrecursorCharge']):\n",
    "    if 1 <= charge <= max_charge_state:\n",
    "        X_charge[i, charge - 1] = 1.0\n",
    "\n",
    "# concatenate peptide sequences and charge states to create the final input (X)\n",
    "X = [np.concatenate((X_peptide[i], np.tile(X_charge[i], (max_sequence_length, 1))), axis=1) for i in range(len(X_peptide))]\n",
    "Y = distinct_df['IonMobility'].values\n",
    "weights = distinct_df['Ms1ProfileCorr'].values\n",
    "\n",
    "# Split the data into training and testing sets\n",
    "X_train, X_test, Y_train, Y_test, weights_train, weights_test = train_test_split(X, Y, weights, test_size=0.1, random_state=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15d97866-e2e3-44cc-b712-803c2af7bcfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def weighted_mean_absolute_error(y_true, y_pred, weights):\n",
    "    \"\"\"\n",
    "    Weighted Mean Absolute Error custom loss function.\n",
    "\n",
    "    :param y_true: True labels\n",
    "    :param y_pred: Predicted labels\n",
    "    :param weights: Weights for each sample (based on Ms1ProfileCorr)\n",
    "    :return: Weighted MAE\n",
    "    \"\"\"\n",
    "    # Calculate the absolute error\n",
    "    error = tf.abs(y_pred - y_true)\n",
    "    transformed_weights = weights+0.1\n",
    "\n",
    "    # Apply weights\n",
    "    weighted_error = error * transformed_weights\n",
    "    \n",
    "    # Return the mean of the weighted error\n",
    "    return tf.reduce_mean(weighted_error)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e019071c-1084-467b-8e0c-47f5f550b98f",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = tf.keras.Sequential([\n",
    "    tf.keras.layers.Flatten(input_shape=(max_sequence_length, num_amino_acids + max_charge_state)),\n",
    "    tf.keras.layers.Dense(300, activation='swish'),\n",
    "    tf.keras.layers.Dense(220, activation='swish'),\n",
    "    tf.keras.layers.Dense(160, activation='swish'),\n",
    "    tf.keras.layers.Dense(100, activation='swish'),\n",
    "    tf.keras.layers.Dense(50, activation='swish'),\n",
    "    tf.keras.layers.Dense(10, activation='swish'),\n",
    "    tf.keras.layers.Dense(1, activation='linear') \n",
    "])\n",
    "\n",
    "# compile\n",
    "model.compile(optimizer='adam', loss=lambda y_true, y_pred: weighted_mean_absolute_error(y_true, y_pred, weights))\n",
    "\n",
    "# train\n",
    "model.fit(np.array(X_train), Y_train, epochs=10, batch_size=32, validation_split=0.1)\n",
    "\n",
    "# predict\n",
    "Y_pred = model.predict(np.array(X_test))\n",
    "Y_pred = Y_pred.flatten()\n",
    "\n",
    "# evaluate the model\n",
    "mse = np.mean((Y_pred - Y_test) ** 2)\n",
    "print(\"Mean Squared Error:\", mse)\n",
    "\n",
    "mae = np.mean(np.abs(Y_pred - Y_test))\n",
    "print(\"Mean Absolute Error:\", mae)\n",
    "\n",
    "med_ae = np.median(np.abs(Y_pred - Y_test))\n",
    "print(\"Median Absolute Error:\", med_ae)\n",
    "\n",
    "coefficient_of_dermination = r2_score(Y_test, Y_pred)\n",
    "print(\"R-squared: \",coefficient_of_dermination)\n",
    "\n",
    "model.save(f'IM_model')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302f3628-8eb9-41a3-a733-c64675d03e95",
   "metadata": {},
   "outputs": [],
   "source": [
    "# apply the model to predict the full FASTA-digested proteome\n",
    "ori_path = 'mTRAQ_CI_H.tsv'\n",
    "df_ori = pd.read_csv(ori_path, delimiter='\\t')\n",
    "df_ori2 = pd.DataFrame(df_ori)\n",
    "df_ori2 = df_ori2[~df_ori2['PeptideSequence'].str.contains('X|U')] #remove peptides with X or U amino-acids\n",
    "distinct_df = df_ori2.drop_duplicates(subset='transition_group_id')\n",
    "\n",
    "df_ori2['fragment_charge_series'] = df_ori2['FragmentType'].astype(str) + \"_\"+df_ori2['FragmentCharge'].astype(str) +\"_\"+ df_ori2['FragmentSeriesNumber'].astype(str)\n",
    "pivot_df_ori = df_ori2.pivot_table(index=['transition_group_id','PeptideSequence','PrecursorCharge'], columns='fragment_charge_series', values='LibraryIntensity', aggfunc='first')\n",
    "\n",
    "# Resetting the index to flatten the DataFrame\n",
    "pivot_df_ori.reset_index(inplace=True)\n",
    "\n",
    "part1 = ['b', 'y']  # First part (b or y)\n",
    "part2 = [1, 2]      # Second part (1 or 2)\n",
    "part3 = range(3, 25)  # Third part (3 through 25)\n",
    "\n",
    "# Generate all combinations\n",
    "all_combinations = itertools.product(part1, part2, part3)\n",
    "# Create column names from combinations\n",
    "all_columns = ['_'.join(map(str, comb)) for comb in all_combinations]\n",
    "additional_columns = ['transition_group_id', 'PeptideSequence', 'PrecursorCharge']  # Replace with your actual column names\n",
    "all_columns.extend(additional_columns)\n",
    "#print(all_columns)\n",
    "\n",
    "df_reindexed_ori = pivot_df_ori.reindex(columns=all_columns)\n",
    "df_reindexed_ori = df_reindexed_ori.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39ad98c3-38f5-47f8-b1a2-582a0e1e84f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_peptide = [one_hot_encode_sequence(seq) for seq in df_reindexed_ori['PeptideSequence']]\n",
    "X_charge = np.zeros((len(df_reindexed_ori), max_charge_state), dtype=np.float32)\n",
    "\n",
    "for i, charge in enumerate(df_reindexed_ori['PrecursorCharge']):\n",
    "    if 1 <= charge <= max_charge_state:\n",
    "        X_charge[i, charge - 1] = 1.0\n",
    "\n",
    "# Concatenate peptide sequences and charge states to create the final input (X)\n",
    "X = [np.concatenate((X_peptide[i], np.tile(X_charge[i], (max_sequence_length, 1))), axis=1) for i in range(len(X_peptide))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ecf0e4f-e67f-4e50-915e-b5ef3f10ea3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = model.predict(np.array(X))\n",
    "\n",
    "all_preds_df = pd.DataFrame(predictions)\n",
    "lim_df = df_reindexed_ori['transition_group_id']\n",
    "\n",
    "combined_df = pd.concat([lim_df.reset_index(drop=True), all_preds_df.reset_index(drop=True)], axis=1)\n",
    "joined_pred = pd.merge(df_ori2, combined_df, on='transition_group_id', how='inner')\n",
    "\n",
    "# save the library. Now it has updated IM's.\n",
    "joined_pred.to_csv('mTRAQ_FASTA_DIANNPred_IM_only.tsv', sep='\\t', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
